vpr/src/device/rr_graph_obj.h

```diff
@@ -0,0 +1,647 @@
+
+    /* Get coordinate of a node. (xlow, xhigh, ylow, yhigh):
+     *   For OPIN/IPIN/SOURCE/SINK, xlow = xhigh and ylow = yhigh
+     *   FIXME: should confirm if this is still the case when a logic block has a height > 1
+     *   For CHANX/CHANY, (xlow, ylow) and (xhigh, yhigh) represent
+     *   where the routing segment starts and ends.
+     *   Note that our convention alway keeps
+     *   xlow <= xhigh and ylow <= yhigh
+     *   Therefore, (xlow, ylow) is a starting point for a CHANX/CHANY in INC_DIRECTION
+     *   (xhigh, yhigh) is a starting point for a CHANX/CHANY in INC_DIRECTION
+     *                           
+     *   Example :
+     *   CHANX in INC_DIRECTION  
+     *   (xlow, ylow)                (xhigh, yhigh)
+     *        |                           |
+     *       \|/                         \|/
+     *        ---------------------------->
+     *
+     *   CHANX in DEC_DIRECTION  
+     *   (xlow, ylow)                (xhigh, yhigh)
+     *        |                           |
+     *       \|/                         \|/
+     *        <----------------------------
+     *
+     *   CHANY in INC_DIRECTION
+     *       
+     *      /|\ <-------(xhigh, yhigh)
+     *       |
+     *       |  <-------(xlow, ylow)
+     *
+     *   CHANY in DEC_DIRECTION
+     *       
+     *       |  <-------(xhigh, yhigh)
+     *       |
+     *      \|/ <-------(xlow, ylow)
+     */
+    short node_xlow(const RRNodeId& node) const;
+    short node_ylow(const RRNodeId& node) const;
+    short node_xhigh(const RRNodeId& node) const;
+    short node_yhigh(const RRNodeId& node) const;
+    /* Get the length of a routing track. 
+     * Note that it is ONLY meaningful for CHANX and CHANY, which represents the number of logic blocks that a routing track spans
+     * For nodes that are OPIN/IPIN/SOURCE/SINK, the length is supposed to be always 1
+     */
+    short node_length(const RRNodeId& node) const;
+    /* A short-cut function to get coordinates of a node */
+    vtr::Rect<short> node_bounding_box(const RRNodeId& node) const;
+
+    /* Get node starting and ending points in routing channels. 
+     * See details in the figure for bounding box 
+     * Applicable to routing track nodes only!!! 
+     */
+    vtr::Point<short> node_start_coordinate(const RRNodeId& node) const;
+    vtr::Point<short> node_end_coordinate(const RRNodeId& node) const;
+
+    /* Get the capacity of a node. 
+     * Literally, how many nets can be mapped to the node. 
+     * Typically, each node has a capacity of 1 but special nodes (SOURCE and SINK) will have a 
+     * large number due to logic equivalent pins
+     * See Vaughn Betz's book for more details
+     */
+    short node_capacity(const RRNodeId& node) const;
+    short node_fan_in(const RRNodeId& node) const;
+    short node_fan_out(const RRNodeId& node) const;
+
+    /* Get the ptc_num of a node
+     * ptc number is a feature number of each node for indexing by node type
```

Comment:
I'd say
The ptc (pin, track, or class) number is an integer that allows you to differentiate between wires, pins or sources/sinks with overlapping x,y coordinates or extent.  This is useful for drawing rr-graphs nicely. For example, all the CHANX rr_nodes that have the same y coordinate and x-coordinates that overlap will have different ptc numbers, by convention starting at 0. This allows them to be drawn without overlapping, as the ptc num can be used as a track identifier. The main routing code does not care about ptc num.